services/telemetry/telemetry.go

```diff
@@ -234,3 +381,79 @@ func getLinuxDistribution(procVersion string) string {
 	}
 	return "unknown"
 }
+
+func (s *Service) runRetries(ctx context.Context) {
+	l.Debug("Start telemetry retries goroutine")
+
+	go func() {
+		for {
+			select {
+			case task := <-s.ch:
+				if err := s.waitAndRetry(ctx, task); err != nil {
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+}
+
+func (s *Service) waitAndRetry(ctx context.Context, task *retryTask) error {
+	d := time.Until(task.t)
+	t := time.NewTimer(d)
+	defer t.Stop()
+
+	select {
+	case <-t.C:
+		s.retry(ctx, task)
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+
+	return nil
+}
+
+func (s *Service) retry(ctx context.Context, task *retryTask) {
+	rCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	l.Debugf("Retry send telemetry request: %s", task.req)
+	err := s.sendV2Request(rCtx, task.req)
+	if err != nil {
+		l.Debugf("sendV2Request: %+v", err)
+		if task.cnt >= retryCnt {
+			l.Debugf("Retry count exceeded, limit: %d, req: %s", retryCnt, task.req)
+			return
+		}
+
+		task.cnt++
+		task.t = time.Now().Add(retryBackoff)
+
+		s.tryToPushToQueue(task)
+	}
+}
+
+func (s *Service) tryToPushToQueue(task *retryTask) bool {
+	select {
+	case s.ch <- task:
+		return true
+	default:
+		l.Debugf("Retry queue overflow, throw away request: %s.", task.req)
+		return false
+	}
+}
+
+func (s *Service) queueToRetry(req *reporterv1.ReportRequest) {
+	s.tryToPushToQueue(&retryTask{
+		cnt: 1, // nolint:mnd
+		t:   time.Now().Add(retryBackoff),
+		req: req,
+	})
+}
+
+type retryTask struct {
+	cnt int32
+	t   time.Time
+	req *reporterv1.ReportRequest
+}
```

Comment:
Overall, it seems to be too complicated for that task.

I think that starting a separate goroutine for each event that tries sending it X times with increasing intervals until success or until all X attempts are made would be much easier and good enough given our numbers.

Reply:
I assumed that for linear backoff queue is a good solution as events always naturally sorted and it allows spawn a single goroutine 🤔

Reply:
How many elements we will have in a queue? Zero or one :) It is a good general solution, but in that case it is hardly needed

Reply:
The same for goroutines: it is good not to spawn goroutines without limits – but we have a natural limit of _one_ goroutine in that task.

Reply:
Done